File: skills-mcp-server/tests/test_sync_script.py
"""
Tests for Python sync-skills.py script

Tests the Python implementation directly:
- Script syntax and structure
- Required functions exist
- Script can be executed
"""

import pytest
import subprocess
import sys
from pathlib import Path


@pytest.fixture
def sync_script():
    """Path to the Python sync script."""
    root = Path(__file__).parent.parent
    script = root / "src" / "resources" / "sync-skills.py"
    assert script.exists(), f"Sync script not found: {script}"
    return script


class TestScriptStructure:
    """Test the structure and syntax of the sync script."""
    
    def test_script_has_valid_syntax(self, sync_script):
        """Test that the script has valid Python syntax."""
        with open(sync_script) as f:
            code = f.read()
        
        try:
            compile(code, str(sync_script), 'exec')
        except SyntaxError as e:
            pytest.fail(f"Script has syntax error: {e}")
    
    def test_script_has_shebang(self, sync_script):
        """Test that the script has proper shebang line."""
        with open(sync_script) as f:
            first_line = f.readline()
        
        assert first_line.startswith('#!/usr/bin/env python3'), \
            "Script should start with python3 shebang"
    
    def test_script_has_required_functions(self, sync_script):
        """Test that all required functions are present."""
        with open(sync_script) as f:
            content = f.read()
        
        required_functions = [
            'check_git_installed',
            'read_repo_list',
            'clone_or_pull_repo',
            'scan_all_skills',
            'update_agents_md',
            'main',
        ]
        
        for func in required_functions:
            assert f'def {func}(' in content, \
                f"Missing required function: {func}"
    
    def test_script_has_proper_imports(self, sync_script):
        """Test that the script imports required modules."""
        with open(sync_script) as f:
            content = f.read()
        
        required_imports = [
            'import os',
            'import subprocess',
            'import sys',
            'from pathlib import Path',
            'import frontmatter',
        ]
        
        for imp in required_imports:
            assert imp in content, f"Missing import: {imp}"
    
    def test_script_size_is_reasonable(self, sync_script):
        """Test that the script size is within expected range."""
        size = sync_script.stat().st_size
        assert 9000 < size < 12000, \
            f"Script size {size} bytes seems unusual (expected ~10KB)"


class TestScriptExecution:
    """Test script execution (requires git to be installed)."""
    
    def test_script_can_be_compiled(self, sync_script):
        """Test that Python can compile the script."""
        result = subprocess.run(
            [sys.executable, '-m', 'py_compile', str(sync_script)],
            capture_output=True,
            text=True,
        )
        
        assert result.returncode == 0, \
            f"Script compilation failed: {result.stderr}"
    
    def test_script_shows_help_or_runs(self, sync_script):
        """Test that script can be invoked (may fail if git not installed)."""
        # Just test that the script can be invoked without syntax errors
        # Don't actually run it as it would modify files
        result = subprocess.run(
            [sys.executable, '-c', f'import sys; exec(open("{sync_script}").read())'],
            capture_output=True,
            text=True,
            timeout=5,
        )
        
        # Script may fail due to missing git or other dependencies
        # but should not have syntax errors
        if result.returncode != 0:
            # Check it's not a syntax error
            assert 'SyntaxError' not in result.stderr, \
                f"Script has syntax error: {result.stderr}"


# Run with: pytest tests/test_sync_script.py -v

File: skills-mcp-server/tests/test-sync-script.ts
#!/usr/bin/env tsx
/**
 * Tests for TypeScript sync-skills.ts script
 * 
 * Tests the TypeScript implementation directly:
 * - Script syntax and structure
 * - Required functions exist
 * - Script can be compiled
 */

import * as fs from 'fs';
import * as path from 'path';
import { fileURLToPath } from 'url';
import { execSync } from 'child_process';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let testsRun = 0;
let testsPassed = 0;
let testsFailed = 0;

function assert(condition: boolean, message: string) {
  if (!condition) {
    throw new Error(`Assertion failed: ${message}`);
  }
}

async function test(name: string, fn: () => void | Promise<void>) {
  testsRun++;
  try {
    await fn();
    testsPassed++;
    console.log(`  ✅ ${name}`);
  } catch (error: any) {
    testsFailed++;
    console.log(`  ❌ ${name}`);
    console.log(`     Error: ${error.message}`);
  }
}

async function runTests() {
  console.log('\n🧪 Testing TypeScript sync-skills.ts Script\n');
  console.log('='.repeat(70));

  const syncScriptPath = path.join(__dirname, '..', 'src', 'resources', 'sync-skills.ts');
  
  // Verify script exists
  if (!fs.existsSync(syncScriptPath)) {
    console.error(`❌ Script not found: ${syncScriptPath}`);
    process.exit(1);
  }

  const scriptContent = fs.readFileSync(syncScriptPath, 'utf-8');

  // Test Suite 1: Script Structure
  console.log('📋 Test Suite 1: Script Structure');
  console.log('-'.repeat(70));

  await test('Script has valid TypeScript syntax', () => {
    try {
      execSync(`npx tsc --noEmit --skipLibCheck "${syncScriptPath}"`, {
        cwd: path.join(__dirname, '..'),
        stdio: 'pipe',
      });
    } catch (error: any) {
      throw new Error(`TypeScript compilation failed: ${error.stderr?.toString() || error.message}`);
    }
  });

  await test('Script has proper shebang line', () => {
    const firstLine = scriptContent.split('\n')[0];
    assert(firstLine.startsWith('#!/usr/bin/env node'), 'Script should start with node shebang');
  });

  await test('Script has required functions', () => {
    const requiredFunctions = [
      'readRepoList',
      'scanAllSkills',
      'updateAgentsMd',
      'main',
    ];

    for (const func of requiredFunctions) {
      assert(
        scriptContent.includes(`function ${func}(`),
        `Missing required function: ${func}`
      );
    }
  });

  await test('Script has proper imports', () => {
    const requiredImports = [
      "import * as fs from 'fs'",
      "import * as path from 'path'",
      "import * as child_process from 'child_process'",
    ];

    for (const imp of requiredImports) {
      assert(scriptContent.includes(imp), `Missing import: ${imp}`);
    }
  });

  await test('Script size is reasonable', () => {
    const size = scriptContent.length;
    assert(size > 9000 && size < 13000, `Script size ${size} bytes seems unusual (expected ~10-11KB)`);
  });

  // Test Suite 2: Script Content
  console.log('\n📄 Test Suite 2: Script Content');
  console.log('-'.repeat(70));

  await test('Script has configuration constants', () => {
    assert(scriptContent.includes('SCRIPT_DIR'), 'Missing SCRIPT_DIR constant');
    assert(scriptContent.includes('SKILLS_DIR'), 'Missing SKILLS_DIR constant');
    assert(scriptContent.includes('CONFIG_FILE'), 'Missing CONFIG_FILE constant');
    assert(scriptContent.includes('AGENTS_MD'), 'Missing AGENTS_MD constant');
  });

  await test('Script has proper error handling', () => {
    assert(scriptContent.includes('try'), 'Script should have try-catch blocks');
    assert(scriptContent.includes('catch'), 'Script should have try-catch blocks');
  });

  await test('Script has main execution block', () => {
    assert(scriptContent.includes('main()'), 'Script should call main()');
  });

  await test('Script handles frontmatter parsing', () => {
    assert(
      scriptContent.includes('frontmatter') || scriptContent.includes('parseFrontmatter'),
      'Script should handle frontmatter parsing'
    );
  });

  // Test Suite 3: Script Compilation
  console.log('\n🔨 Test Suite 3: Script Compilation');
  console.log('-'.repeat(70));

  await test('Script compiles to JavaScript', () => {
    try {
      const tempDir = path.join(__dirname, '.test-compile');
      if (!fs.existsSync(tempDir)) {
        fs.mkdirSync(tempDir);
      }

      execSync(
        `npx tsc "${syncScriptPath}" --outDir "${tempDir}" --skipLibCheck`,
        {
          cwd: path.join(__dirname, '..'),
          stdio: 'pipe',
        }
      );

      const compiledFile = path.join(tempDir, 'sync-skills.js');
      assert(fs.existsSync(compiledFile), 'Compiled file should exist');

      // Cleanup
      fs.rmSync(tempDir, { recursive: true });
    } catch (error: any) {
      throw new Error(`Compilation failed: ${error.stderr?.toString() || error.message}`);
    }
  });

  await test('Compiled script has valid JavaScript', () => {
    try {
      const tempDir = path.join(__dirname, '.test-compile');
      if (!fs.existsSync(tempDir)) {
        fs.mkdirSync(tempDir);
      }

      execSync(
        `npx tsc "${syncScriptPath}" --outDir "${tempDir}" --skipLibCheck`,
        {
          cwd: path.join(__dirname, '..'),
          stdio: 'pipe',
        }
      );

      const compiledFile = path.join(tempDir, 'sync-skills.js');
      const compiledContent = fs.readFileSync(compiledFile, 'utf-8');
      
      // Check for valid JavaScript patterns
      assert(compiledContent.includes('function'), 'Compiled code should have functions');
      assert(compiledContent.includes('exports') || compiledContent.includes('export'), 'Compiled code should have exports');

      // Cleanup
      fs.rmSync(tempDir, { recursive: true });
    } catch (error: any) {
      throw new Error(`JavaScript validation failed: ${error.message}`);
    }
  });

  // Summary
  console.log('\n' + '='.repeat(70));
  console.log('📊 Test Summary\n');
  console.log(`  Total:  ${testsRun}`);
  console.log(`  ✅ Passed: ${testsPassed}`);
  console.log(`  ❌ Failed: ${testsFailed}`);
  console.log();

  if (testsFailed > 0) {
    console.log('❌ Some tests failed!\n');
    process.exit(1);
  } else {
    console.log('✅ All tests passed!\n');
    process.exit(0);
  }
}

runTests().catch((error) => {
  console.error('\n💥 Test runner failed:', error);
  process.exit(1);
});

